proposals/p2875.md

```diff
@@ -0,0 +1,469 @@
+function call operation, and how callable entities such as the name of a
+parameterized class fit into the picture -- are they functions or something
+else?
+
+## Background
+
+This proposal assumes the reader understands how functions and function pointers
+work in C or C++.
+
+## Proposal
+
+A function declaration in Carbon introduces a new, unique type, called a
+_function type_. The function name is a stateless value of that function type.
+
+Function calls use C-like syntax: an expression naming a callable is followed by
+a syntax resembling a tuple of arguments. There are several kinds of callable:
+
+-   Functions, and more generally values of function types.
+-   Bound member functions, such as `my_vector.Begin`.
+-   Lambdas, once they are introduced to Carbon.
+-   Parameterized entities, such as a generic class `Vector` or a generic
+    interface `AddWith`.
+-   Values of dependent types that are constrained to be callable.
+-   User-defined class types that overload function call syntax.
+
+In the first four cases, argument deduction is used to determine the values of
+deduced arguments, and then pattern matching is used to bind parameter values to
+arguments.
+
+In the remaining cases, the built-in `Call` interface is used to determine the
+semantics of the call, and the call `F(args)` is translated into
+`F.(Call(ArgTypes).Op)(args)`. Note that this is itself a bound member function
+call, and `Call(ArgTypes)` is a call to a parameterized entity, so this
+transformation is only performed once.
+
+## Details
+
+### Functions and function types
+
+A function declaration such as
+
+```
+fn F[T:! type](x: T) -> T { return x; }
+```
+
+introduces a value named `F`, whose type is a unique _function type_. Distinct
+functions have distinct function types, even if they have the same signature. A
+function type is an empty, trivial type. There is no way to name a function type
+other than asking for the type of the function value.
+
+```
+// Compile-time function.
+fn TypeOf[T:! type](x: T) -> type { return T; }
+
+// ✅ `F` is a first-class value with a first-class type.
+alias FType = TypeOf(F);
+var my_f: FType = F;
+```
+
+Function values are regular values that can be stored in variables, passed to
+functions, and so on.
+
+```
+fn G() -> i32 {
+  // ✅ `my_f` has function type `FType`. This is a direct call to `F`.
+  return my_f(1);
+}
+
+fn Sort[template F:! type, T:! type](v: Vector(T)*, f: F);
+fn Compare(a: i32, b: i32) -> Ordering { return a.(Ordered.Compare)(b); }
+fn SortInts(v: Vector(i32)*) { Sort(v, Compare); }
+```
+
+For the purpose of the orphan rule, a function type is considered to be declared
+by the function declaration that introduces the function value.
+
+### Call syntax
+
+Calls take the form `a(b, c, d)`. This is syntactically equivalent to a primary
+expression followed by a tuple literal, except that a trailing comma is not
+required for the case of a single argument, `a(b)`.
```

Comment:
```suggestion
required for the case of a single argument, either `a(b)` or `a(b,)`.
```

Just looking for a different way to emphasize the meaning here (to your point about tuple syntax, most of my concern is that this'll be copied into the design, and the direct comparison won't be obvious to a new reader).

Reply:
I've taken another pass over this section; I think this new approach sufficiently emphasizes what I'm going for here and should still be clear when copied into the design.